From 98476dc8b23d3ac7cb7e818a066f519c888ef8ac Mon Sep 17 00:00:00 2001
From: Emil Velikov <emil.velikov@collabora.com>
Date: Thu, 7 Jun 2018 16:24:45 +0100
Subject: [PATCH] shell-completion: systemctl: pass current word to all
 list_unit*
Bug: https://github.com/systemd/systemd/issues/7185
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/bionic/+source/systemd/+bug/1529152
Origin: upstream, https://github.com/systemd/systemd/pull/8863

Earlier patch added the current word to the performance critical paths.
Here we add it to every place, for consistency sake.

Suggested-by: Yu Watanabe (yuwata)
Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>
---
 shell-completion/bash/systemctl.in | 18 +++++++++---------
 shell-completion/zsh/_systemctl.in |  6 +++---
 2 files changed, 12 insertions(+), 12 deletions(-)

--- a/shell-completion/bash/systemctl.in
+++ b/shell-completion/bash/systemctl.in
@@ -79,7 +79,7 @@ __filter_units_by_properties () {
         done
 }
 
-__get_all_units      () { { __systemctl $1 list-unit-files; __systemctl $1 list-units --all; } \
+__get_all_units      () { { __systemctl $1 list-unit-files "$2*"; __systemctl $1 list-units --all "$2*"; } \
         | { while read -r a b; do [[ $a =~ @\. ]] || echo " $a"; done; }; }
 __get_template_names () { __systemctl $1 list-unit-files "$2*" \
         | { while read -r a b; do [[ $a =~ @\. ]] && echo " ${a%%@.*}@"; done; }; }
@@ -111,7 +111,7 @@ __get_disabled_units () { __systemctl $1
         | { while read -r a b c  ; do [[ $b == "disabled" ]] && echo " $a"; done; }; }
 __get_masked_units   () { __systemctl $1 list-unit-files "$2*" \
         | { while read -r a b c  ; do [[ $b == "masked"   ]] && echo " $a"; done; }; }
-__get_all_unit_files () { { __systemctl $1 list-unit-files; } | { while read -r a b; do echo " $a"; done; }; }
+__get_all_unit_files () { { __systemctl $1 list-unit-files "$2*"; } | { while read -r a b; do echo " $a"; done; }; }
 
 __get_machines() {
         local a b
@@ -227,7 +227,7 @@ _systemctl () {
                 comps="${VERBS[*]}"
 
         elif __contains_word "$verb" ${VERBS[ALL_UNITS]}; then
-                comps=$( __get_all_units $mode )
+                comps=$( __get_all_units $mode "$cur" )
                 compopt -o filenames
 
         elif __contains_word "$verb" ${VERBS[ENABLED_UNITS]}; then
@@ -267,7 +267,7 @@ _systemctl () {
 
         elif __contains_word "$verb" ${VERBS[ISOLATABLE_UNITS]}; then
                 comps=$( __filter_units_by_property $mode AllowIsolate yes \
-                      $( __get_all_units $mode ) )
+                      $( __get_all_units $mode "$cur" ) )
                 compopt -o filenames
 
         elif __contains_word "$verb" ${VERBS[FAILED_UNITS]}; then
@@ -281,10 +281,10 @@ _systemctl () {
         elif __contains_word "$verb" ${VERBS[TARGET_AND_UNITS]}; then
                 if __contains_word "$prev" ${VERBS[TARGET_AND_UNITS]} \
                 || __contains_word "$prev" ${OPTS[STANDALONE]}; then
-                        comps=$( __systemctl $mode list-unit-files --type target --all \
+                        comps=$( __systemctl $mode list-unit-files --type target --all "$cur*" \
                         | { while read -r a b; do echo " $a"; done; } )
                 else
-                        comps=$( __get_all_unit_files $mode )
+                        comps=$( __get_all_unit_files $mode "$cur" )
                 fi
                 compopt -o filenames
 
@@ -304,7 +304,7 @@ _systemctl () {
                 compopt -o filenames
 
         elif __contains_word "$verb" ${VERBS[TARGETS]}; then
-                comps=$( __systemctl $mode list-unit-files --type target --full --all \
+                comps=$( __systemctl $mode list-unit-files --type target --full --all "$cur*" \
                         | { while read -r a b; do echo " $a"; done; } )
         fi
 
